src/modes/run.ts

```diff
@@ -1,5 +1,340 @@
+
+        _(`Found ${events.length} events`);
+
+        // process the events
+        for (const event of events) {
+          if (plan[event.blockNumber] === undefined) {
+            plan[event.blockNumber] = new Set();
+          }
+
+          plan[event.blockNumber].add(event);
+        }
+
+        // only possible string value for toBlock is 'latest'
+        if (typeof toBlock === "number") {
+          fromBlock = toBlock + 1;
+        }
+      } while (toBlock !== "latest" && toBlock !== currentBlockNumber);
+
+      const block = await provider.getBlock(currentBlockNumber);
+
+      // Replay only the blocks that had some events.
+      for (const [blockNumber, events] of Object.entries(plan)) {
+        _(`[run_rebuild] Processing block ${blockNumber}`);
+        try {
+          await processBlock(
+            this,
+            Number(blockNumber),
+            events,
+            block.number,
+            block.timestamp
+          );
+
+          // Set the last processed block to this iteration's block number
+          this.registry.lastProcessedBlock = Number(blockNumber);
+          await this.registry.write();
+        } catch {
+          console.error(
+            `[warmUp:chainId:${chainId}] Error processing block ${blockNumber}`
+          );
+        }
+        _(`Block ${blockNumber} has been processed.`);
+      }
+
+      // Set the last processed block to the current block number
+      this.registry.lastProcessedBlock = currentBlockNumber;
+
+      // Save the registry
+      await this.registry.write();
+
+      // It may have taken some time to process the blocks, so refresh the current block number
+      // and check if we are in sync
+      currentBlockNumber = await provider.getBlockNumber();
+
+      // If we are in sync, let it be known
+      if (currentBlockNumber === this.registry.lastProcessedBlock) {
+        this.inSync = true;
+      } else {
+        // Otherwise, we need to keep processing blocks
+        this.inSync = false;
+        fromBlock = this.registry.lastProcessedBlock + 1;
+        plan = {};
+      }
+    } while (!this.inSync);
+
+    _(oneShot ? "Chain watcher is in sync" : "Chain watcher is warmed up");
+    _(`Last processed block: ${this.registry.lastProcessedBlock}`);
+
+    // Notifying that the chain watcher is in sync
+    this.inSync = true;
+
+    // If one-shot, return
+    if (oneShot) {
+      return;
+    }
+
+    // Otherwise, run the block watcher
+    return await this.runBlockWatcher();
+  }
+
+  /**
+   * Run the block watcher for the chain. As new blocks come in:
+   * 1. Check if there are any `ConditionalOrderCreated` events, and index these.
+   * 2. Check if any orders want to create discrete orders.
+   */
+  public async runBlockWatcher() {
+    const { provider, chainId } = this.chainContext;
+    const _ = (s: string) =>
+      console.log(`[runBlockWatcher:chainId:${chainId}] ${s}`);
+    // Watch for new blocks
+    _("Subscribe to new blocks");
```

Comment:
👀

Reply:
Heh, your inquisitiveness is tracked in #60